firebase-dataconnect/src/main/kotlin/com/google/firebase/dataconnect/serializers/TimestampSerializer.kt

```diff
@@ -56,41 +56,50 @@ public object TimestampSerializer : KSerializer<Timestamp> {
     return "$serializedSecond.${serializedNano}Z"
   }
 
-  /**
-   * Note: Timestamp sent from SDK is always 9 digits nanosecond precision, meaning there are 9
-   * digits in SSSSSSSSS parts. However, when running against different databases, this precision
-   * might change, and server will truncate it to 0/3/6 digits precision without throwing an error.
-   */
+  // TODO: Replace this implementation with Instant.parse() once minSdkVersion is bumped to at
+  //  least 26 (Build.VERSION_CODES.O).
   private fun timestampFromString(str: String): Timestamp {
-    // TODO: support +/- time zone offsets as well (b/337299540)
-    require(str.uppercase().endsWith("Z")) {
-      "the last character should be 'Z', but got '${str.takeLast(1)}' (str=$str)"
+    val time = str.uppercase()
+
+    val timeZoneOffsetFormatRegex =
+      Regex("^\\d{4}-\\d{2}-\\d{2}T\\d{2}:\\d{2}:\\d{2}(\\.\\d{0,9})?Z$")
+    val nanoFormatRegex = Regex("^\\d{4}-\\d{2}-\\d{2}T\\d{2}:\\d{2}:\\d{2}[+-]\\d{2}:\\d{2}$")
+
+    require(time.matches(timeZoneOffsetFormatRegex) || time.matches(nanoFormatRegex)) {
+      "str=$time cannot be deserialized into Timestamp."
     }
 
-    // TODO: Replace this implementation with Instant.parse() once minSdkVersion is bumped to at
-    //  least 26 (Build.VERSION_CODES.O).
     val position = ParsePosition(0)
     val seconds = run {
-      val date = dateFormatter.parse(str.uppercase(), position)
+      val date = dateFormatter.parse(time.uppercase(), position)
       requireNotNull(date)
       require(position.index == 19) {
-        "position.index=${position.index}, but expected 19 (str=$str)"
+        "position.index=${position.index}, but expected 19 (str=$time)"
       }
       Timestamp(date).seconds
     }
 
-    val nanoseconds =
-      if (str[position.index] == 'Z' || str[position.index] == 'z') {
-        0
-      } else {
-        require(str[position.index] == '.') {
-          "str[${position.index}]=='${str[position.index]}', but expected '.' (str=$str)"
-        }
-        val nanosecondsStr = str.substring(position.index + 1, str.length - 1)
-        nanosecondsStr.padEnd(9, '0').toInt()
-      }
+    /** When Timestamp sent from server is in time zone offset format. */
```

Comment:
I'm not sure what this comment means. Can it just be removed? If you leave it here and reword it, please use the inline comment syntax `//` rather than the javadoc comment `/** */`.